### DISCUSSION TITLE
Find a dummy string for "unused eslint-disable directive"

### DISCUSSION THREAD

**Discussion by Splines** [eslint/eslint#18114] [Q&A] [answered]
I've recently written an [**eslint-plugin-erb**](https://github.com/Splines/eslint-plugin-erb) to lint JavaScript inside ERB files (embedded Ruby syntax). Autocorrecting looks like this:
![erb lint showcase](https://private-user-images.githubusercontent.com/37160523/293907467-623d6007-b4f5-41ce-be76-5bc0208ed636.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NzczNzYsIm5iZiI6MTcwNzc3NzA3NiwicGF0aCI6Ii8zNzE2MDUyMy8yOTM5MDc0NjctNjIzZDYwMDctYjRmNS00MWNlLWJlNzYtNWJjMDIwOGVkNjM2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDIyMzExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlYzU2NmE5OGY2ODgyNjI0NzFjN2MyMThmMmY3YTNkZmRhODAwNWRjODU2YTg3MzgxMTgzNGI4ZjI2MmI5MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LT5v9DQWf6DW5SdYYkUJxTfLCV2yRFZQ0FPQ5riu57M)

Let's consider this code snippet:
```js
    const a = "before"
<%= render partial: 'questions/domains/expression',
            locals: { a: a, b: b,
                       c: "c" } %>
 console.log("afterwards");
```

Every ruby comment starts with `<%` and ends with `%>`. So my idea for the plugin was: why not simply replace all ruby strings with a dummy string such that the file can still be linted by ESLint. So, the above snippet gets translated into:

```js
    const a = "before"
/* eslint-disable */{}/* eslint-enable */
 console.log("afterwards");
```

I keep track of how my replacement changed the line and column numbers. This is such that after retrieving the warnings/errors/fixes by ESLint, I can map the numbers back to their original position in the file. So far so good. This is actually working very well in a real world codebase.

The problem I now encounter is the choice of the dummy string `/* eslint-disable */{}/* eslint-enable */`. I've chosen it such that the output produced is still valid JS and such that we don't get any syntax errors when passing the transformed file to ESLint. The `eslint-disable` is necessary because we want to ignore any errors introduced by the dummy string itself. And it works fine.

But there is one real world use case where this is breaking for us:

```js
var posted = <%= @posted %>;
// The plugin converts this line into:
var posted = /* eslint-disable */{}/* eslint-enable */;
```

However, the `{}` is a valid empty object that gets assigned to the `posted` variable. That's why ESLint is -- legimately -- warning about `Unused eslint-disable directive` as there is no warning produced by `{}` in this case. And worse: the autocorrect feature does not do its job correctly in this case. I don't want to disable this "unused eslint-disable directive" warning for the entire codebase, that's why I'm asking:

💡 **Do you know / can come up with any other dummy string that could cover all use cases**, i.e. a string that produces valid JS but where ESLint also has some complaints about, such that we don't get the "Unused eslint-disable-directive" warning? Thanks for any input 😇

**Comment by Rec0iL99** [comment]
> I've chosen it such that the output produced is still valid JS and such that we don't get any syntax errors when passing the transformed file to ESLint

Hi @Splines, using `/* eslint-disable */` does not make ESLint ignore syntax errors. You would still need to pass valid JS to ESLint for the file to be parsed successfully. I think choosing any dummy comment other than `eslint-disable` should work for your specific use case.

**Reply by Splines** [reply]
You're right that `/* eslint-disable */` does not ignore syntax errors, that's why I use `/* eslint-disable */{}/* eslint-enable */` as dummy string which produces valid JS (note the `{}` inside the string). The problem is not that I get syntax errors, but that I get the described warning. The goal is to find a string that produces valid JS syntax once dropped in for the `<%` `%>` code, but still induces an ESLint *linting* warning/error that then gets ignored due to `eslint-disable`.

**Reply by Splines** [reply]
As an example, see the one I mentioned above:

```js
    const a = "before"
<%= render partial: 'questions/domains/expression',
            locals: { a: a, b: b,
                       c: "c" } %>
 console.log("afterwards");
```

This works fine with my plugin as ESLint is able to parse the following string:
```js
    const a = "before"
/* eslint-disable */{}/* eslint-enable */
 console.log("afterwards");
```

It will then output some linting warinngs/errors, e.g. there are unnecessary whitespaces before `const a` that it will remove. The warning/error that `{}` is solely on its own line will be ignored due to the `eslint-disable`. ESLint then goes on with linting the last line. This works. But the situation is different for:

```js
var posted = <%= @posted %>;
```

which gets translated in a preprocessing step to:

```js
var posted = /* eslint-disable */{}/* eslint-enable */;
```

Here, ESLint analyzing the string `var posted = {};` will not output any errors/warnings related to `{}`, hence the user of my plugin will get a warning that we use the `eslint-disble` "special comment" even though it is not necessary in this case. That's of course really unexpected for the user who might not even know what kind of magic the plugin performs with this `eslint-disable` preprocessing.

That's the problem why I need to find another dummy string that works in these cases as well, i.e. a dummy string that results in syntactically correct JavaScript but not correct JavaScript *linting-wise* (e.g. `{}` on an empty line will produce linting errors, which is good for us). I hope this explanation got my point across ;)

**Reply by Splines** [reply]
> I think choosing any dummy comment other than eslint-disable should work for your specific use case.

I don't see how any dummy comment other than eslint-disable would work in this case, e.g. choose the dummy comment `YEAH`. Then I pass the following string to ESLint:

```js
var posted = YEAH;
```

And of course ESLint will complain about `YEAH` probably with something like "undefined...". The user of the plugin will therefore get an error on the following line:

```js
var posted = <%= @posted %>;
```

telling them that "YEAH" is not defined. Of course, the user is then really surprised what their code even has to do with "YEAH" 😅

**Reply by Rec0iL99** [reply, marked as answer]
I meant a dummy block comment `/* YEAH */`, just like `/* eslint-disable */`, because then you wouldn't have to worry about `reportUnusedDisableDirectives`. 

```js
var posted = /* YEAH */ {} /* YEAH */;
```

But, in your case, since you want ESLint to ignore linting for a portion of the file, I think the only way to move forward is to continue using `/* eslint-disable */` and disable the `reportUnusedDisableDirectives` option.

**Reply by Splines** [reply]
Thanks, I will suggest the users of my plugin to add the following to their configuration ;)

```js
linterOptions: {
      // The "unused disable directive" is set to "warn" by default.
      // For the ERB plugin to work correctly, you must disable
      // this directive to avoid issues described here
      // https://github.com/eslint/eslint/discussions/18114
      // If you're using the CLI, you might also use the following flag:
      // --report-unused-disable-directives-severity=off
      reportUnusedDisableDirectives: "off",
},
// your other configuration options
```
